lib/spack/spack/architecture.py

```diff
@@ -56,49 +56,197 @@
+        if self.micro_architecture.vendor == 'generic':
+            return str(self)
+
+        return syaml_dict([
+            ('name', str(self.micro_architecture.name)),
+            ('vendor', str(self.micro_architecture.vendor)),
+            ('features', sorted(
+                str(x) for x in self.micro_architecture.features
+            )),
+            ('generation', self.micro_architecture.generation),
+            ('parents', [str(x) for x in self.micro_architecture.parents])
+        ])
+
+    def isa_target_for(self, compiler):
+        """Returns the name to be used to optimize for the current target
+        using the compiler passed as argument, or None if the information
+        is not available.
+
+        Args:
+            compiler: object with a ``name`` and ``version`` attribute for
+                the compiler being used.
+
+        Raises:
+            UnsupportedMicroArchitecture: if the compiler version passed as
+                argument doesn't support the current micro-architecture.
+        """
+        # No information available on compiler: return None
+        if compiler.name not in self.micro_architecture.compilers:
+            return None
+
+        # Get the information on supported compiler versions
+        constraints = self.micro_architecture.compilers[compiler.name]
+        if not isinstance(constraints, Sequence):
+            constraints = [constraints]
+
+        for constraint in constraints:
+            version_request = spack.version.ver(constraint['versions'])
+            if compiler.version.satisfies(version_request):
+                # The name to be used defaults to the micro-architecture name,
+                # but can be overridden by compiler info in the JSON file
+                name = constraint.get('name', None) or self.name
+                return name
+
+        msg = 'cannot produce optimized binary for micro-architecture "{0}"' \
+            ' with {1.name}@{1.version!s} ' \
+            '[supported compiler versions are {2}]'
+        msg = msg.format(self.micro_architecture.name, compiler,
+                         ', '.join([x['versions'] for x in constraints]))
+        raise UnsupportedMicroArchitecture(msg)
 
     def __repr__(self):
-        return self.__str__()
+        cls_name = self.__class__.__name__
+        fmt = cls_name + '({0}, {1})'
+        return fmt.format(repr(self.micro_architecture),
+                          repr(self.module_name))
 
     def __str__(self):
-        return self.name
+        return str(self.micro_architecture)
+
+    def __contains__(self, cpu_flag):
+
+        def sse3_proxy():
+            return any(
+                x in self.micro_architecture for x in ('sse3', 'ssse3')
+            )
+
+        def avx512_proxy():
+            avx512_flags = [
+                'avx512f', 'avx512vl', 'avx512bw', 'avx512dq', 'avx512cd',
+            ]
+            return any(x in self.micro_architecture for x in avx512_flags)
```

Comment:
> See my suggestion above (in cpu.py) about pushing all flags from ancestors into the list for every node -- then we do not need this proxy stuff, do we?

We might still need it. For instance `avx512` above means generically any of the avx 512 instruction and that won't be covered by pushing flags down to children. I like though the idea to push aliases to the JSON file. In case, would it be fine with you to add enclosing properties to the current JSON? What I mean is something like:
```json
{
  "microarchitectures": {
    "x86_64": {},
    "...": {}
  },
  "query_aliases": {
    "avx512": "anyOf(avx512', avx512vl, avx512bw, avx512dq, avx512cd)"
  }
}
```

Reply:
💯 This is exactly what I think it should look like.  Minor nitpick: `anyOf` -> `any_of` (just for consistency).  Would it make sense to call these `feature aliases` instead of `query_aliases`?  There is no concept of a query in the JSON file (and I think that's good).

Reply:
Further nitpick: the alias shouldn't be a string.  Shouldn't it just be:

```json
"avx512": {
    "any_of": ["avx512", "avx512vl", "avx512bw", ...]
}
```

That way we don't have to parse the parsed json.

Reply:
We might even expose feature aliases for CPUs that don't expose them in cpuinfo, e.g.:
```json
"altivec": {
  "families": ["ppc64", "ppc64le"]
}
```

Reply:
Yes! I like it.

Reply:
Done in 28b2d96935581b8bfb7b8e3c1c3d39d81412dcbe